src/main/java/org/xbill/DNS/NioUdpClient.java

```diff
@@ -125,29 +127,87 @@ public void processReadyKey(SelectionKey key) {
       buffer.flip();
       byte[] data = new byte[read];
       System.arraycopy(buffer.array(), 0, data, 0, read);
-      verboseLog(
-          "UDP read",
-          channel.socket().getLocalSocketAddress(),
-          channel.socket().getRemoteSocketAddress(),
-          data);
+      verboseLog("UDP read", channel.socket().getLocalSocketAddress(), source, data);
       silentCloseChannel();
-      f.complete(data);
+      f.complete(Collections.singletonList(data));
       pendingTransactions.remove(this);
     }
 
-    private void silentCloseChannel() {
+    void silentCloseChannel() {
       try {
         channel.disconnect();
         channel.close();
       } catch (IOException e) {
         // ignore, we either already have everything we need or can't do anything
       }
     }
+
+    void closeTransaction() {
+      silentCloseChannel();
+      f.completeExceptionally(new SocketTimeoutException("Query timed out"));
+    }
+  }
+
+  private static class MultiAnswerTransaction extends Transaction {
+    MultiAnswerTransaction(
+        byte[] query,
+        int max,
+        long endTime,
+        DatagramChannel channel,
+        SocketAddress remoteSocketAddress,
+        CompletableFuture<List<byte[]>> f) {
+      super(query, max, endTime, channel, remoteSocketAddress, f);
+    }
+
+    public void processReadyKey(SelectionKey key) {
+      if (!key.isReadable()) {
+        silentCloseChannel();
+        f.completeExceptionally(new EOFException("channel not readable"));
+        pendingTransactions.remove(this);
+        return;
+      }
+
+      DatagramChannel channel = (DatagramChannel) key.channel();
+      ByteBuffer buffer = ByteBuffer.allocate(max);
+      SocketAddress source;
+      int read;
+      try {
+        source = channel.receive(buffer);
+        read = buffer.position();
+        if (read <= 0 || source == null) {
+          return; // ignore this datagram
+        }
+      } catch (IOException e) {
+        silentCloseChannel();
+        f.completeExceptionally(e);
+        pendingTransactions.remove(this);
+        return;
+      }
+
+      buffer.flip();
+      byte[] data = new byte[read];
+      System.arraycopy(buffer.array(), 0, data, 0, read);
+      verboseLog("UDP read", channel.socket().getLocalSocketAddress(), source, data);
+      answers.add(data);
+    }
+
+    private ArrayList<byte[]> answers = new ArrayList<>();
+
+    @Override
+    void closeTransaction() {
```

Comment:
closeTransaction is only triggered when the timeout has expired. Is this really the only condition on which you want to return answers? What about the case where one is only interested in the first answer, as fast as possible?

The timeout processing currently only runs every second. Is that still sufficient?